hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterClientProtocol.java

```diff
@@ -1928,9 +1928,14 @@ public BatchedEntries<OpenFileEntry> listOpenFiles(long prevId,
   @Override
   public void msync() throws IOException {
     rpcServer.checkOperation(NameNode.OperationCategory.READ, true);
-    Set<FederationNamespaceInfo> nss = namenodeResolver.getNamespaces();
+    // Only msync to nameservices with observer reads enabled.
+    Set<FederationNamespaceInfo> allNamespaces = namenodeResolver.getNamespaces();
     RemoteMethod method = new RemoteMethod("msync");
-    rpcClient.invokeConcurrent(nss, method);
+    Set<FederationNamespaceInfo> namespacesEligibleForObserverReads = allNamespaces
+        .stream()
+        .filter(ns -> rpcClient.isNamespaceObserverReadEligible(ns.getNameserviceId()))
+        .collect(Collectors.toSet());
+    rpcClient.invokeConcurrent(namespacesEligibleForObserverReads, method);
```

Comment:
If the ```namespacesEligibleForObserverReads``` land up being 0, I think this would throw us an exception, there is check if you go inside
```
    if (locations.isEmpty()) {
      throw new IOException("No remote locations available");
    } 
```

I think if there is no location we should just return here